File: 0299-bulls-and-cows/0299-bulls-and-cows.cpp
class Solution {
public:
    string getHint(string secret, string guess) {
        int bulls = 0, cows = 0;
        vector<int> cntS(10, 0), cntG(10, 0);

        for (int i = 0; i < secret.size(); i++) {
            if (secret[i] == guess[i]) {
                bulls++;
            } else {
                cntS[secret[i] - '0']++;
                cntG[guess[i] - '0']++;
            }
        }

        for (int d = 0; d < 10; d++) {
            cows += min(cntS[d], cntG[d]);
        }

        return to_string(bulls) + "A" + to_string(cows) + "B";
    }
};
// Count bulls directly when digits match at the same position.
// For non-bull digits, store their frequencies in secret and guess.
// Cows are the common digits between these frequency counts.
// Return result in the format "xAyB".

File: 0299-bulls-and-cows/0299-bulls-and-cows.java
class Solution {
    public String getHint(String secret, String guess) {
        int bulls = 0;
        int cows = 0;
        int[] freq = new int[10];
        // Count bulls and store remaining secret digits
        for (int i = 0; i < secret.length(); i++) {
            if (secret.charAt(i) == guess.charAt(i)) {
                bulls++;
            } else {
                freq[secret.charAt(i) - '0']++;
            }
        }
        // Count cows
        for (int i = 0; i < guess.length(); i++) {
            if (secret.charAt(i) != guess.charAt(i)) {
                int digit = guess.charAt(i) - '0';

                if (freq[digit] > 0) {
                    cows++;
                    freq[digit]--;
                }
            }
        }
        return bulls + "A" + cows + "B";
    }
}